---
title: C++异常
author: Sunglow
top: false
cover: false
toc: false
mathjax: false
summary: 'null'
categories:
  - 软件工程
  - Cpp
tags:
  - 编程
  - C++
date: 2018-02-28 21:25:13
keywords:
---

### C++中构造函数和析构函数可以抛出异常吗?  

答：只有构造函数可以。  

析构函数不能抛出异常，处理析构函数异常的正确方式是将异常封装在析构函数内部，而不是抛出异常，如下代码所示。
```
~ClassName(){
    try{
        do_something();
    }
    catch{
        //...
    }
}
```
原因如下：C++异常处理模型有责任处理那些因为出现异常而失效的对象，处理方式是调用这些失效对象的析构函数，释放掉它们占用的资源。如果析构函数再抛出异常，则会陷入无尽递归嵌套之中，因此这是不被允许的。